README
##Usage
To start training run the following commands:
`cd DynamicEnv
python train.py`


DynamicEnv/changes.txt
- added minimal distances of all link to obstacle points as additional observation
- deactivated lidar observation

DynamicEnv/test.py
import os
import sys
import gym
import numpy as np
from stable_baselines3 import PPO
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.noise import NormalActionNoise
from stable_baselines3.common.callbacks import CallbackList, CheckpointCallback, EvalCallback, StopTrainingOnMaxEpisodes
from stable_baselines3.common.utils import set_random_seed
from stable_baselines3.common.monitor import Monitor
from stable_baselines3.common.vec_env import DummyVecEnv, SubprocVecEnv, VecNormalize
from stable_baselines3.common.torch_layers import BaseFeaturesExtractor
from typing import Callable
import torch
import torch.nn as nn
import torch.nn.functional as F

CURRENT_PATH = os.path.abspath(__file__)
sys.path.insert(0,os.path.dirname(CURRENT_PATH))
from env import Env

params = {
    'is_render': True, 
    'is_good_view': True,
    'is_train' : False,
    'show_boundary' : True,
    'add_moving_obstacle' : False,
    'moving_obstacle_speed' : 0.15,
    'moving_init_direction' : -1,
    'moving_init_axis' : 0,
    'workspace' : [-0.4, 0.4, 0.3, 0.7, 0.2, 0.4],
    'max_steps_one_episode' : 1024,
    'num_obstacles' : 3,
    'prob_obstacles' : 0.8,
    'obstacle_box_size' : [0.04,0.04,0.002],
    'obstacle_sphere_radius' : 0.04       
}


if __name__=='__main__':

    env = Env(
        is_render=params['is_render'],
        is_good_view=params['is_good_view'],
        is_train=params['is_train'],
        show_boundary=params['show_boundary'],
        add_moving_obstacle=params['add_moving_obstacle'],
        moving_obstacle_speed=params['moving_obstacle_speed'],
        moving_init_direction=params['moving_init_direction'],
        moving_init_axis=params['moving_init_axis'],
        workspace=params['workspace'],
        max_steps_one_episode=params['max_steps_one_episode'],
        num_obstacles=params['num_obstacles'],
        prob_obstacles=params['prob_obstacles'],
        obstacle_box_size=params['obstacle_box_size'],
        obstacle_sphere_radius=params['obstacle_sphere_radius']
        )
    # load drl model
    model = PPO.load('./models/best_reach_ppo/best_model', env=env)

    while True:
        done = False
        obs = env.reset()
        while not done:
            action, _states = model.predict(obs)
            obs, rewards, done, info = env.step(action)
            # print(info)


DynamicEnv/math_util.py
import numpy as np
import math
# epsilon for testing whether a number is close to zero
_EPS = np.finfo(float).eps * 4.0

# axis sequences for Euler angles
_NEXT_AXIS = [1, 2, 0, 1]

# map axes strings to/from tuples of inner axis, parity, repetition, frame
_AXES2TUPLE = {
    'sxyz': (0, 0, 0, 0), 'sxyx': (0, 0, 1, 0), 'sxzy': (0, 1, 0, 0),
    'sxzx': (0, 1, 1, 0), 'syzx': (1, 0, 0, 0), 'syzy': (1, 0, 1, 0),
    'syxz': (1, 1, 0, 0), 'syxy': (1, 1, 1, 0), 'szxy': (2, 0, 0, 0),
    'szxz': (2, 0, 1, 0), 'szyx': (2, 1, 0, 0), 'szyz': (2, 1, 1, 0),
    'rzyx': (0, 0, 0, 1), 'rxyx': (0, 0, 1, 1), 'ryzx': (0, 1, 0, 1),
    'rxzx': (0, 1, 1, 1), 'rxzy': (1, 0, 0, 1), 'ryzy': (1, 0, 1, 1),
    'rzxy': (1, 1, 0, 1), 'ryxy': (1, 1, 1, 1), 'ryxz': (2, 0, 0, 1),
    'rzxz': (2, 0, 1, 1), 'rxyz': (2, 1, 0, 1), 'rzyz': (2, 1, 1, 1)}

_TUPLE2AXES = dict((v, k) for k, v in _AXES2TUPLE.items())
def euler_from_matrix(matrix, axes='sxyz'):
    """Return Euler angles from rotation matrix for specified axis sequence.

    axes : One of 24 axis sequences as string or encoded tuple

    Note that many Euler angle triplets can describe one matrix.

    >>> R0 = euler_matrix(1, 2, 3, 'syxz')
    >>> al, be, ga = euler_from_matrix(R0, 'syxz')
    >>> R1 = euler_matrix(al, be, ga, 'syxz')
    >>> numpy.allclose(R0, R1)
    True
    >>> angles = (4.0*math.pi) * (numpy.random.random(3) - 0.5)
    >>> for axes in _AXES2TUPLE.keys():
    ...    R0 = euler_matrix(axes=axes, *angles)
    ...    R1 = euler_matrix(axes=axes, *euler_from_matrix(R0, axes))
    ...    if not numpy.allclose(R0, R1): print axes, "failed"

    """
    try:
        firstaxis, parity, repetition, frame = _AXES2TUPLE[axes.lower()]
    except (AttributeError, KeyError):
        _ = _TUPLE2AXES[axes]
        firstaxis, parity, repetition, frame = axes

    i = firstaxis
    j = _NEXT_AXIS[i+parity]
    k = _NEXT_AXIS[i-parity+1]

    M = np.array(matrix, dtype=np.float64, copy=False)[:3, :3]
    if repetition:
        sy = math.sqrt(M[i, j]*M[i, j] + M[i, k]*M[i, k])
        if sy > _EPS:
            ax = math.atan2( M[i, j],  M[i, k])
            ay = math.atan2( sy,       M[i, i])
            az = math.atan2( M[j, i], -M[k, i])
        else:
            ax = math.atan2(-M[j, k],  M[j, j])
            ay = math.atan2( sy,       M[i, i])
            az = 0.0
    else:
        cy = math.sqrt(M[i, i]*M[i, i] + M[j, i]*M[j, i])
        if cy > _EPS:
            ax = math.atan2( M[k, j],  M[k, k])
            ay = math.atan2(-M[k, i],  cy)
            az = math.atan2( M[j, i],  M[i, i])
        else:
            ax = math.atan2(-M[j, k],  M[j, j])
            ay = math.atan2(-M[k, i],  cy)
            az = 0.0

    if parity:
        ax, ay, az = -ax, -ay, -az
    if frame:
        ax, az = az, ax
    return ax, ay, az

def euler_from_quaternion(quaternion, axes='sxyz'):
    """Return Euler angles from quaternion for specified axis sequence.

    >>> angles = euler_from_quaternion([0.06146124, 0, 0, 0.99810947])
    >>> numpy.allclose(angles, [0.123, 0, 0])
    True

    """
    return euler_from_matrix(quaternion_matrix(quaternion), axes)

def quaternion_matrix(quaternion):
    """Return homogeneous rotation matrix from quaternion.

    >>> R = quaternion_matrix([0.06146124, 0, 0, 0.99810947])
    >>> numpy.allclose(R, rotation_matrix(0.123, (1, 0, 0)))
    True

    """
    q = np.array(quaternion[:4], dtype=np.float64, copy=True)
    nq = np.dot(q, q)
    if nq < _EPS:
        return np.identity(4)
    q *= math.sqrt(2.0 / nq)
    q = np.outer(q, q)
    return np.array((
        (1.0-q[1, 1]-q[2, 2],     q[0, 1]-q[2, 3],     q[0, 2]+q[1, 3], 0.0),
        (    q[0, 1]+q[2, 3], 1.0-q[0, 0]-q[2, 2],     q[1, 2]-q[0, 3], 0.0),
        (    q[0, 2]-q[1, 3],     q[1, 2]+q[0, 3], 1.0-q[0, 0]-q[1, 1], 0.0),
        (                0.0,                 0.0,                 0.0, 1.0)
        ), dtype=np.float64)

def get_angle(vector_1, vector_2):
    unit_vector_1 = vector_1 / np.linalg.norm(vector_1)
    unit_vector_2 = vector_2 / np.linalg.norm(vector_2)
    angle=np.arccos(np.dot(vector_1,vector_2)/(np.linalg.norm(vector_1)*np.linalg.norm(vector_2)))
    angle_deg = np.rad2deg(angle)
    return angle, angle_deg

DynamicEnv/pybullet_util.py
import os
import pybullet as p
import pybullet_data as pd
import numpy as np

class MotionExecute():
    def __init__(
        self,
        robot_id, 
        base_link,
        effector_link
        ):
        self.robot_id = robot_id
        self.base_link = base_link
        self.effector_link = effector_link
        
    def getinversePoisition(self, position_desired,orientation_desired=[]):
        joints_info = []
        joint_ll = []
        joint_ul = []
        useOrientation=len(orientation_desired)
        for i in range(self.effector_link+1):
            joints_info.append(p.getJointInfo(self.robot_id, i))
        robotEndEffectorIndex = self.effector_link
        numJoints = p.getNumJoints(self.robot_id)
        useNullSpace = 1
        ikSolver = 1
        pos = [position_desired[0], position_desired[1], position_desired[2]]
        # end effector points down, not up (in case useOrientation==1)
        if useOrientation:
            orn = p.getQuaternionFromEuler([orientation_desired[0],orientation_desired[1] , orientation_desired[2]])
        if (useNullSpace == 1):
            if (useOrientation > 0):
                jointPoses = p.calculateInverseKinematics(self.robot_id, robotEndEffectorIndex, pos, orn)
            else:
                jointPoses = p.calculateInverseKinematics(self.robot_id,
                                                    robotEndEffectorIndex,
                                                    pos,
                                                    lowerLimits=joint_ll,
                                                    upperLimits=joint_ul,
                                                )
                
            # print(jointPoses)
        else:
            if (useOrientation > 0):
                jointPoses = p.calculateInverseKinematics(self.robot_id,
                                                        robotEndEffectorIndex,
                                                        pos,
                                                        orn,
                                                        solver=ikSolver,
                                                        maxNumIterations=100,
                                                        residualThreshold=.01)
                # print(jointPoses)
            else:
                jointPoses = p.calculateInverseKinematics(self.robot_id,
                                                        robotEndEffectorIndex,
                                                        pos,
                                                        solver=ikSolver)
                p.getJointInfo()
        return jointPoses

    def go_to_target(self, position, orientation):
        jointPoses = self.getinversePoisition(position, orientation)
        for i in range(self.base_link, self.effector_link):
            p.resetJointState(bodyUniqueId=self.robot_id,
                                    jointIndex=i,
                                    targetValue=jointPoses[i-1],
                                    )
        return

DynamicEnv/rays_to_indicator.py
import numpy as np
class RaysCauculator():
    def __init__(self, obs_rays):
        self.obs_rays = obs_rays
        self.indicator = np.zeros((10,), dtype=np.int8)

    def get_indicator(self):
        rays_sum = []
        self.obs_tip = self.obs_rays[0:25]
        self.side_1 = self.obs_rays[25:35]
        self.side_2 = self.obs_rays[35:45]
        self.side_3 = self.obs_rays[45:55]
        self.side_4 = self.obs_rays[55:65]
        self.side_5 = self.obs_rays[65:75]
        self.side_6 = self.obs_rays[75:85]
        self.side_7 = self.obs_rays[85:95]
        self.side_8 = self.obs_rays[95:105]
        self.obs_top = self.obs_rays[105:]
        rays_sum.append(self.obs_tip)
        rays_sum.append(self.side_1)
        rays_sum.append(self.side_2)
        rays_sum.append(self.side_3)
        rays_sum.append(self.side_4)
        rays_sum.append(self.side_5)
        rays_sum.append(self.side_6)
        rays_sum.append(self.side_7)
        rays_sum.append(self.side_8)
        rays_sum.append(self.obs_top)
        for i in range(10):
            if rays_sum[i].min()>=0.99:
                self.indicator[i] = 0
            if 0.5<rays_sum[i].min()<0.99:
                self.indicator[i] = 1
            if rays_sum[i].min()<=0.5:
                self.indicator[i] = 2
        return self.indicator

DynamicEnv/train.py
import os
import sys
import gym
import numpy as np
from stable_baselines3 import PPO
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.noise import NormalActionNoise
from stable_baselines3.common.callbacks import CallbackList, CheckpointCallback, EvalCallback, StopTrainingOnMaxEpisodes
from stable_baselines3.common.utils import set_random_seed
from stable_baselines3.common.monitor import Monitor
from stable_baselines3.common.vec_env import DummyVecEnv, SubprocVecEnv, VecNormalize
from stable_baselines3.common.torch_layers import BaseFeaturesExtractor
from typing import Callable
import torch
import torch.nn as nn
import torch.nn.functional as F

CURRENT_PATH = os.path.abspath(__file__)
sys.path.insert(0, os.path.dirname(CURRENT_PATH))
from env import Env

params = {
    'is_render': False,
    'is_good_view': False,
    'is_train' : True,
    'show_boundary' : False,
    'add_moving_obstacle' : False,
    'moving_obstacle_speed' : 0.15,
    'moving_init_direction' : -1,
    'moving_init_axis' : 0,
    'workspace' : [-0.4, 0.4, 0.3, 0.7, 0.2, 0.4],
    'max_steps_one_episode' : 1024,
    'num_obstacles' : 3,
    'prob_obstacles' : 0.8,
    'obstacle_box_size' : [0.04,0.04,0.002],
    'obstacle_sphere_radius' : 0.04       
}

def make_env(rank: int, seed: int = 0) -> Callable:
    """
    Utility function for multiprocessed env.
    
    :param env_id: (str) the environment ID
    :param num_env: (int) the number of environment you wish to have in subprocesses
    :param seed: (int) the inital seed for RNG
    :param rank: (int) index of the subprocess
    :return: (Callable)
    """
    def _init() -> gym.Env:
        env = Env(
            is_render=params['is_render'],
            is_good_view=params['is_good_view'],
            is_train=params['is_train'],
            show_boundary=params['show_boundary'],
            add_moving_obstacle=params['add_moving_obstacle'],
            moving_obstacle_speed=params['moving_obstacle_speed'],
            moving_init_direction=params['moving_init_direction'],
            moving_init_axis=params['moving_init_axis'],
            workspace=params['workspace'],
            max_steps_one_episode=params['max_steps_one_episode'],
            num_obstacles=params['num_obstacles'],
            prob_obstacles=params['prob_obstacles'],
            obstacle_box_size=params['obstacle_box_size'],
            obstacle_sphere_radius=params['obstacle_sphere_radius']
            )
        env = Monitor(env)
        env.seed(seed + rank)
        return env
    set_random_seed(seed)
    return _init

if __name__=='__main__':

    # Separate evaluation env
    eval_env = Env(
        is_render=params['is_render'],
        is_good_view=params['is_good_view'],
        is_train=False,
        show_boundary=params['show_boundary'],
        add_moving_obstacle=params['add_moving_obstacle'],
        moving_obstacle_speed=params['moving_obstacle_speed'],
        moving_init_direction=params['moving_init_direction'],
        moving_init_axis=params['moving_init_axis'],
        workspace=params['workspace'],
        max_steps_one_episode=params['max_steps_one_episode'],
        num_obstacles=params['num_obstacles'],
        prob_obstacles=params['prob_obstacles'],
        obstacle_box_size=params['obstacle_box_size'],
        obstacle_sphere_radius=params['obstacle_sphere_radius']
        )
    eval_env = Monitor(eval_env)
    # load env
    env = SubprocVecEnv([make_env(i) for i in range(8)])
    # Stops training when the model reaches the maximum number of episodes
    callback_max_episodes = StopTrainingOnMaxEpisodes(max_episodes=1e8, verbose=1)

    # Use deterministic actions for evaluation
    eval_callback = EvalCallback(eval_env, best_model_save_path='./models/best_reach_ppo/',
                       log_path='./models/best_reach_ppo/', eval_freq=5120,
                       deterministic=True, render=False)
    
    # Save a checkpoint every ? steps
    checkpoint_callback = CheckpointCallback(save_freq=5120, save_path='./models/reach_ppo_ckp_logs/',
                                        name_prefix='reach')
    # Create the callback list
    callback = CallbackList([checkpoint_callback, callback_max_episodes, eval_callback])
    model = PPO("MultiInputPolicy", env, batch_size=256, verbose=1, tensorboard_log='./models/reach_ppo_tf_logs/')  # create random model
    # model = PPO.load('./models/reach_ppo_ckp_logs/reach_1024000_steps', env=env) # load pretrained model
    model.learn(
        total_timesteps=1e10,
        n_eval_episodes=64,
        callback=callback,
        reset_num_timesteps=False)  # False when you do not want to reset timesteps to 0

    model.save('./models/reach_ppo')
